File: DAA LAB/1st program merge sort.c
#include <stdio.h>
#include <stdlib.h>

int count = 0;

void merge(int a[], int l, int m, int r)
{
    int i = l, j = m + 1, k = l, b[r + 1];
    while (i <= m && j <= r)
    {
        count++;
        if (a[i] < a[j])
            b[k++] = a[i++];
        else
            b[k++] = a[j++];
    }
    while (i <= m)
        b[k++] = a[i++];
    while (j <= r)
        b[k++] = a[j++];
    for (i = l; i <= r; i++)
        a[i] = b[i];
}

void merge_sort(int a[], int l, int r)
{

    if (l < r)
    {
        count++;
        int m = (l + r) / 2;
        merge_sort(a, l, m);
        merge_sort(a, m + 1, r);
        merge(a, l, m, r);
    }
}

int main()
{
    int n;
    printf("Enter the size of the array: ");
    scanf("%d", &n);
    int a[n], i;
    printf("Enter %d elements: ", n);
    for (i = 0; i < n; i++)
        scanf("%d", &a[i]);
    merge_sort(a, 0, n - 1);
    printf("The Merge sorted array is:");
    for (i = 0; i < n; i++)
        printf("\t%d", a[i]);
    printf("\nBasic operation executed is: %d times\n", count);
    printf("Table of number of times basic operation gets executed:\n");
    printf("n\tasc\tdes\trand\n");
    int c1, c2;
    for (i = 2; i <= 1024; i *= 2)
    {
        int a[i], b[i], c[i], j;
        for (j = 0; j < i; j++)
        {
            a[j] = j;
            b[j] = i - j;
            c[j] = rand();
        }
        count = 0;
        merge_sort(a, 0, i - 1);
        c1 = count;
        count = 0;
        merge_sort(b, 0, i - 1);
        c2 = count;
        count = 0;
        merge_sort(c, 0, i - 1);
        printf("%d\t%d\t%d\t%d\n", i, c1, c2, count);
    }
    return 0;
}


/*

output of the program:

Enter the size of the array: 10
Enter 10 elements: 1 4 51 65498 1 5187  54 616 465 156 
The Merge sorted array is:      1       1       4       51      54      156     465     616     5187    65498
Basic operation executed is: 33 times
Table of number of times basic operation gets executed:
n       asc     des     rand
2       2       2       2
4       7       7       8
8       19      19      24
16      47      47      58
32      111     111     146
64      255     255     369
128     575     575     863
256     1279    1279    1976
512     2815    2815    4495
1024    6143    6143    9973


Enter the size of the array: 5
Enter 5 elements: 5 4 3 2 1
The Merge sorted array is:      1       2       3       4       5
Basic operation executed is: 9 times
Table of number of times basic operation gets executed:
n       asc     des     rand
2       2       2       2
4       7       7       8
8       19      19      24
16      47      47      58
32      111     111     146
64      255     255     369
128     575     575     863
256     1279    1279    1976
512     2815    2815    4495
1024    6143    6143    9973


*/

File: DAA LAB/1st program merge sort.java
import java.util.Scanner;
import java.util.Random;

public class MergeSort {

    static int count=0;

    public static void main(String[] args) {
        Scanner sc=new Scanner(System.in);
        System.out.print("Enter the number of elements: ");
        int n=sc.nextInt();
        int[] a=new int[n];
        System.out.print("Enter "+n+" elements: ");
        int i,j;
        for(i=0;i<n;i++)
            a[i]=sc.nextInt();
        mergeSort(a,0,n-1);
        System.out.print("Merge sorted array is:");
        for(int var:a)
            System.out.print(" "+var);
        System.out.println("\nNumber of times basic operation executed is: "+count);
        int c1,c2;
        System.out.println("Time complexity table is");
        System.out.println("n\ta\tr\td");
        for(i=1;i<=4096;i*=2){
            int[] asc=new int[i];
            int[] rand=new int[i];
            int[] des=new int[i];
            Random random=new Random(i);
            for(j=0;j<i;j++)
            {
                asc[j]=j+1;
                rand[j]=random.nextInt();
                des[j]=i-j;
            }
            count=0;
            mergeSort(asc,0,i-1);
            c1=count;
            count=0;
            mergeSort(rand,0,i-1);
            c2=count;
            count=0;
            mergeSort(des,0,i-1);
            System.out.println(i+"\t"+c1+"\t"+c2+"\t"+count);
        }
    }

    public static void mergeSort(int[] a,int l,int r)
    {
        if(l<r){
            count++;
            int m=l+(r-l)/2;
            mergeSort(a,l,m);
            mergeSort(a,m+1,r);
            conquer(a,l,m,r);
        }
    }

    static void conquer(int[] a,int l,int m,int r)
    {
        int i=l,j=m+1,k=l;
        int[] b=new int[r+1];
        while(i<=m&&j<=r)
        {
            count++;
            if(a[i]<a[j])
                b[k++]=a[i++];
            else
                b[k++]=a[j++];
        }
        while(i<=m)
            b[k++]=a[i++];
        while(j<=r)
            b[k++]=a[j++];
        for(i=l;i<=r;i++)
            a[i]=b[i];
    }
}


/*

Output of the code:

Enter the number of elements: 4
Enter 4 elements: 4 1 3 2
Merge sorted array is: 1 2 3 4
Number of times basic operation executed is: 8
Time complexity table is
n	a	r	d
1	0	0	0
2	2	2	2
4	7	8	7
8	19	22	19
16	47	61	47
32	111	160	111
64	255	370	255
128	575	872	575
256	1279	1982	1279
512	2815	4475	2815
1024	6143	9974	6143
2048	13311	22003	13311
4096	28671	48073	28671



Enter the number of elements: 10
Enter 10 elements: 1 54 51 651 165 61 74 9846 1 49
Merge sorted array is: 1 1 49 51 54 61 74 165 651 9846
Number of times basic operation executed is: 31
Time complexity table is
n	a	r	d
1	0	0	0
2	2	2	2
4	7	8	7
8	19	22	19
16	47	61	47
32	111	160	111
64	255	370	255
128	575	872	575
256	1279	1982	1279
512	2815	4475	2815
1024	6143	9974	6143
2048	13311	22003	13311
4096	28671	48073	28671

*/
